llvm/lib/Transforms/Scalar/GVN.cpp

```diff
@@ -1324,15 +1390,11 @@ void GVNPass::AnalyzeLoadAvailability(LoadInst *Load, LoadDepVect &Deps,
       continue;
     }
 
-    if (!DepInfo.isLocal()) {
-      UnavailableBlocks.push_back(DepBB);
-      continue;
-    }
```

Comment:
How important is this?  Do you know how `!local` the case you care about is.  My knowledge is a little weak here but I ask because if this is known `isNonLocal` then that suggests you don't need to scan DepBB and can jump straight to the predecessors, which might reduce the amount of refactoring and thus make my earlier comments redundant.